File: gateway.go
package deezer

import (
   "encoding/json"
   "fmt"
   "io/ioutil"
   "net/http"
   "net/http/cookiejar"
   "net/url"
   "strings"
)

func gateway(sngId, apiKey string) error {
   jar, err := cookiejar.New(nil)
   if err != nil {
      return err
   }
   http.DefaultClient.Jar = jar
   // GET
   val := url.Values{}
   req, err := http.NewRequest("GET", API.String(), nil)
   if err != nil {
      return err
   }
   val.Set("api_version", "1.0")
   val.Set("api_token", "null")
   val.Set("method", "deezer.ping")
   req.URL.RawQuery = val.Encode()
   res, err := http.DefaultClient.Do(req)
   if err != nil {
      return err
   }
   defer res.Body.Close()
   // JSON
   body, err := ioutil.ReadAll(res.Body)
   if err != nil {
      return err
   }
   var ping struct {
      Results struct {
         Session string
      }
   }
   json.Unmarshal(body, &ping)
   // POST
   req, err = http.NewRequest(
      "POST",
      "https://api.deezer.com/1.0/gateway.php",
      strings.NewReader(fmt.Sprintf(`{"SNG_ID": %v}`, sngId)),
   )
   if err != nil {
      return err
   }
   val.Set("api_key", apiKey)
   val.Set("method", "song_getData")
   val.Set("output", "3")
   val.Set("input", "3")
   println(ping.Results.Session)
   val.Set("sid", ping.Results.Session)
   req.URL.RawQuery = val.Encode()
   res, err = http.DefaultClient.Do(req)
   if err != nil {
      return err
   }
   // JSON
   body, err = ioutil.ReadAll(res.Body)
   if err != nil {
      return err
   }
   // memcpy(md5, find(page, "PUID\":\""), 32);
   fmt.Printf("%s\n", body)
   return nil
}

File: docs/2021-03-13/gateway.py
from urllib import parse
import os, requests
session = requests.Session()

PRIVATE_API_URL = 'https://www.deezer.com/ajax/gw-light.php?'
args = {
   'api_version': '1.0',
   'api_token': 'null',
   'method': 'deezer.ping'
}
res = session.get(PRIVATE_API_URL + parse.urlencode(args))

PUBLIC_API_URL = 'https://api.deezer.com/1.0/gateway.php?'
args = {
   'api_key': os.environ['DZR_API'],
   'method': 'song_getData',
   'input': '3',
   'output': '3',
   'sid': res.json()['results']['SESSION']
}
res = session.post(PUBLIC_API_URL + parse.urlencode(args), data='{"sng_id":75498415}')
print(res.text)
